Spring 3 auth (#284)
Azure-Samples/ms-identity-java-spring-tutorial: 3-Authorization-II/3-App-CallAPI/README.md

---
page_type: sample
name: Enable your Java Spring Boot daemon to call a protected Web API with its own identity
description: Enable your Java Spring Boot daemon to call a protected Web API with its own identity
languages:
 - java
products:
 - azure-active-directory
 - msal-java
urlFragment: ms-identity-java-spring-tutorial
extensions:
- services: ms-identity
- platform: Java
- endpoint: AAD v2.0
- level: 200
- client: Java Spring Daemon
- service: Java Spring Web API
---

# Enable your Java Spring Boot daemon to call a protected Web API with its own identity

> [![Build status](https://identitydivision.visualstudio.com/IDDP/_apis/build/status/AAD%20Samples/.NET%20client%20samples/ASP.NET%20Core%20Web%20App%20tutorial)](https://identitydivision.visualstudio.com/IDDP/_build/latest?definitionId=XXX)

Reviewer: ```suggestion
```

* [Overview](#overview)
* [Scenario](#scenario)
* [Prerequisites](#prerequisites)
* [Setup the sample](#setup-the-sample)
* [Explore the sample](#explore-the-sample)
* [Troubleshooting](#troubleshooting)
* [How to deploy this sample to Azure](#how-to-deploy-this-sample-to-azure)
* [Next Steps](#next-steps)
* [Contributing](#contributing)
* [Learn More](#learn-more)

## Overview

> This sample demonstrates a Java Spring Daemon calling a Java Spring Web API that is secured using Azure AD.

Reviewer: Does it use MSAL Java -if so, we should mention here (and possibly in the title)

## Scenario

> This sample demonstrates a Java Spring Daemon calling a Java Spring Web API that is secured using Azure AD.

Reviewer: ```suggestion
```

1. Still on the same app registration, select the **App roles** blade to the left.
1. Select **Create app role**:
    1. For **Display name**, enter a suitable name for your application permission, for instance **ToDoList.Read.All**.
    1. For **Allowed member types**, choose **Application** to ensure other applications can be granted this permission.
    1. For **Value**, enter **ToDoList.Read.All**.
    1. For **Description**, enter *Allow the app to read every user's ToDo list using the 'java-spring-webapi-auth'*.
    1. Select **Apply** to save your changes.

    > Repeat the steps above for another app permission named **ToDoList.ReadWrite.All**

##### Configure Optional Claims

1. Still on the same app registration, select the **Token configuration** blade to the left.
1. Select **Add optional claim**:
    1. Select **optional claim type**, then choose **Access**.

>      1. Select the optional claim **idtyp**.

Reviewer: ```suggestion
    1. Select the optional claim **idtyp**.
```

1. Open a terminal or the integrated VSCode terminal.
1. In the root directory as this project, run `cd 3-Authorization-II\3-App-CallAPI\resource-api`.
1. run `mvn clean compile spring-boot:run`.
1. Open a terminal or the integrated VSCode terminal.
1. In the root directory as this project, run `cd 3-Authorization-II\3-App-CallAPI\daemon`.
1. run `mvn clean compile spring-boot:run`.

## Explore the sample
![Experience](./AppCreationScripts/ReadmeFiles/app.png)

- Note the output printed in the console log
- Each line of output is a different API call to your protected API
## Troubleshooting

<details>
	<summary>Expand for troubleshooting info</summary>

Use [Stack Overflow](http://stackoverflow.com/questions/tagged/msal) to get support from the community. Ask your questions on Stack Overflow first and browse existing issues to see if someone has asked your question before.
Make sure that your questions or comments are tagged with [`azure-active-directory` `msal-java` `ms-identity` `msal`].
If you find a bug in the sample, raise the issue on [GitHub Issues](../../../issues).

</details>

### Obtain an Access Token for your App using Client Credentials

> As a daemon application, this sample creates a confidential client application to obtain the access token used to call your protected Web APIs. The confidential client application is configured in [OAuthClientConfiguration](".daemon\src\main\java\com\microsoft\azuresamples\msal4j\configuration\OAuthClientConfiguration.java") class.

Reviewer: ```suggestion
As a daemon application, this sample creates a **confidential client application** to obtain an access token that is used to call your protected web API. The confidential client application is configured in the [OAuthClientConfiguration](".daemon\src\main\java\com\microsoft\azuresamples\msal4j\configuration\OAuthClientConfiguration.java") class.
```

> In this class, we define a ClientRegistration object using information obtained defined in the application.yml file. The ClientRegistration represents a client registered with OAuth2 and holds all the information pertaining to the client

Reviewer: ```suggestion
In this class, we define a `ClientRegistration` object using the parameters defined in the [application.yml](./daemon/resources/application.yml) file. The `ClientRegistration` object represents a client registered with OAuth2 and holds all the information pertaining to the client.
```

```java
    @Bean
    ClientRegistration ADClientRegistration(
            @Value("${spring.security.oauth2.client.provider.azure.token-uri}") String token_uri,
            @Value("${spring.security.oauth2.client.registration.azure.client-id}") String client_id,
            @Value("${spring.security.oauth2.client.registration.azure.client-secret}") String client_secret,
            @Value("${spring.security.oauth2.client.registration.azure.scope}") String scope,
            @Value("${spring.security.oauth2.client.registration.azure.authorization-grant-type}") String authorizationGrantType,
            @Value("${spring.security.oauth2.client.registration.azure.client-authentication-method}") String authMethod
    ) {
        return ClientRegistration
                .withRegistrationId("azure")
                .tokenUri(token_uri)
                .clientId(client_id)
                .clientSecret(client_secret)
                .scope(scope)
                .authorizationGrantType(new AuthorizationGrantType(authorizationGrantType))
                .clientAuthenticationMethod(new ClientAuthenticationMethod(authMethod))
                .build();
    }
```

> We use this client registration to create an AuthorizedClientServiceOAuth2AuthorizedClientManager object, which is used by our daemon application to obtain an OAuth2AuthorizedClient.

Reviewer: ```suggestion
We use this client registration to create an `AuthorizedClientServiceOAuth2AuthorizedClientManager` object, which is used by our daemon application to instantiate `OAuth2AuthorizedClient`.
```

 ```java
    	OAuth2AuthorizeRequest authorizeRequest = OAuth2AuthorizeRequest.withClientRegistrationId("azure")
				.principal("Test Tenant")
				.build();
		OAuth2AuthorizedClient authorizedClient = this.authorizedClientServiceAndManager.authorize(authorizeRequest);
```

The OAuth2AuthorizedClient represents a successfully authorized client that houses the access token and refresh token needed to call our protected Web APIs.
```java
		OAuth2AccessToken accessToken = Objects.requireNonNull(authorizedClient).getAccessToken();
		String accessTokenValue = accessToken.getTokenValue();
        
    	final WebClient apiClient = WebClient.builder()
                .baseUrl(apiAddress)
                .defaultHeader("Authorization", String.format("Bearer %s", accessTokenValue))
                        .build();
```

> ### Validate your Azure access tokens using routes with AADDelegatingOAuth2TokenValidator

Reviewer: ```suggestion
### Validate your Azure access tokens using AADDelegatingOAuth2TokenValidator
```

> While `AADWebSecurityConfigurationAdapter` can be used to protect your routes so only valid users can access it, Azure protected API's require additional claims validation to ensure only valid users are calling you routes. For this, see this app's [AppSecurityConfig](.resource-api/src/main/java/com/microsoft/azuresamples/msal4j/msidentityspringbootwebapi/AppSecurityConfig.java) class.

Reviewer: ```suggestion
Azure protected web API's must perform both signature and claim validation on the incoming access tokens to authorize a call. For extended claim validation, refer to this web API's [AppSecurityConfig](.resource-api/src/main/java/com/microsoft/azuresamples/msal4j/msidentityspringbootwebapi/AppSecurityConfig.java) class.
```

```

1. If you have not made a Java SE Web App in your subscription for this project, select the option, `create`, be pressing enter
1. When prompted for an OS option, select `Linux` by pressing enter.
1. When prompted for a javaVersion, select `Java 11` by entering the associated number for it
1. When prompted for a Pricing Tier, select `B1` or any other option for your pricing tier by entering the associated number for it
1. Finally, press enter on the last prompt to confirm your selections.
1. Open your pom.xml, it should now be updated with build information for deploying to Azure App Services

#### Step 2: Publish your files using Maven (java-spring-webapi-auth)

1. In the terminal window, enter the following command:

```console
        mvn clean package azure-webapp:deploy
```

1. The terminal log should display a successful build after a few minutes stating that the deployment was successful

> :note: The maven plugin uses information from your subscription to deploy to Azure App Services. Deployment is subject to the limitations of your subscription. Modify your pom.xml to meet your subscription limits.

## Next Steps

Learn how to:

> >* [Enable your Java Spring Boot web app to sign in users with Azure Active Directory and call a protected Web api](https://github.com/Azure-Samples/ms-identity-java-spring-tutorial/tree/spring-3-Auth/3-Authorization-II/2-User-CallAPI)

Reviewer: ```suggestion
* [Enable your Java Spring Boot web app to sign in users with Azure Active Directory and call a protected Web api](https://github.com/Azure-Samples/ms-identity-java-spring-tutorial/tree/spring-3-Auth/3-Authorization-II/2-User-CallAPI)
```
